api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirEncounterDaoImpl.java

```diff
@@ -44,4 +57,121 @@ public Encounter getEncounterByUuid(String uuid) {
 		        .createAlias("patient", "p").createAlias("p.identifiers", "pi").add(eq("pi.identifier", patientIdentifier))
 		        .list();
 	}
+	
+	@Override
+	public Collection<Encounter> searchForEncounters(DateRangeParam date, ReferenceParam location,
+	        ReferenceParam participant, ReferenceParam subject) {
+		
+		Criteria criteria = sessionFactory.getCurrentSession().createCriteria(Encounter.class);
+		handleDateRange("encounterDatetime", date).ifPresent(criteria::add);
+		handleLocationReference(criteria, location);
+		handleParticipantReference(criteria, participant);
+		handleSubjectReference(criteria, subject);
+		//TODO: handle ReferenceParam for participant, subject, location
+		
+		return criteria.list();
+	}
+	
+	private void handleLocationReference(Criteria criteria, ReferenceParam locationReference) {
+		if (locationReference != null) {
+			criteria.createAlias("location", "l");
+			
+			if (locationReference.getChain() != null) {
+				switch (locationReference.getChain()) {
+					case Location.SP_ADDRESS_CITY:
+						criteria.add(ilike("l.cityVillage", locationReference.getValue(), MatchMode.START));
+						break;
+					case Location.SP_ADDRESS_STATE:
+						criteria.add(ilike("l.stateProvince", locationReference.getValue(), MatchMode.START));
+						break;
+					case Location.SP_ADDRESS_POSTALCODE:
+						criteria.add(ilike("l.postalCode", locationReference.getValue(), MatchMode.START));
+						break;
+					case Location.SP_ADDRESS_COUNTRY:
+						criteria.add(ilike("l.country", locationReference.getValue(), MatchMode.START));
+						break;
+					case "":
+						criteria.add(eq("l.uuid", locationReference.getValue()));
+						break;
+				}
+			}
+		}
+	}
+	
+	private void handleParticipantReference(Criteria criteria, ReferenceParam participantReference) {
+		if (participantReference != null) {
+			criteria.createAlias("encounterProviders", "ep");
+			
+			if (participantReference.getChain() != null) {
+				switch (participantReference.getChain()) {
+					case Patient.SP_IDENTIFIER:
```

Comment:
Should we be using `Patient.SP_IDENTIFIER` here?

Reply:
@ibacher I guess we should use Practitioner

Reply:
Yes, exactly!